### PR TITLE
MINIFICPP-1426 MSVC: Use /std:c++latest /permissive- and drop VS 2017

### PR DESCRIPTION
**Pull request by szaszm**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [x] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE file?
- [x] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nanofi/examples/CMakeLists.txt [hunk 1]
```diff
@@ -19,57 +19,35 @@
...
+    target_link_libraries(generate_flow nanofi Threads::Threads)
 
-#add_executable(hash_file hash_file.c)
+    target_link_libraries(terminate_handler nanofi Threads::Threads)
 
-#arget_link_libraries(hash_file nanofi Threads::Threads )
+    #add_executable(hash_file hash_file.c)
 
-add_executable(transmit_flow transmit_flow.c)
+    #arget_link_libraries(hash_file nanofi Threads::Threads )
```
Comment (lordgamez) [inline nanofi/examples/CMakeLists.txt:37]: (old code) Can we add a comment why this is commented out?

**Review by lordgamez — APPROVED** [review]
LGTM, thanks!

File: nanofi/examples/CMakeLists.txt (hunk 1 above)
Comment (szaszm) [reply nanofi/examples/CMakeLists.txt:37]: I don't know the reason, just fixed the indentation here after changing the compiler flags above.

Comment (szaszm) [reply nanofi/examples/CMakeLists.txt:37]: It looks like the whole executable is commented out for whatever reason.

Comment (lordgamez) [reply nanofi/examples/CMakeLists.txt:37]: We could write a Jira ticket for it to investigate and fix the issue later.

Comment (szaszm) [reply nanofi/examples/CMakeLists.txt:37]: I don't really see the point of further pursuing this as long as nanofi is barely maintained.
This is where hash_file was introduced: #499 
And this is where it was removed: #514 [MINIFICPP-767 Fix issue with hash](https://issues.apache.org/jira/browse/MINIFICPP-767)

Comment (lordgamez) [reply nanofi/examples/CMakeLists.txt:37]: Yeah, that's okay then, thanks!

File: cmake/BundledAzureSdkCpp.cmake [hunk 2]
```diff
@@ -16,8 +16,13 @@
...
 
 function(use_bundled_libazure SOURCE_DIR BINARY_DIR)
-    set(PATCH_FILE "${SOURCE_DIR}/thirdparty/azure-sdk-cpp-for-cpp/azure-sdk-for-cpp-old-compiler.patch")
-    set(PC "${Patch_EXECUTABLE}" -R -p1 -s -f --dry-run -i "${PATCH_FILE}" || "${Patch_EXECUTABLE}" -p1 -i "${PATCH_FILE}")
+    set(PATCH_FILE1 "${SOURCE_DIR}/thirdparty/azure-sdk-cpp-for-cpp/azure-sdk-for-cpp-old-compiler.patch")
+    set(PATCH_FILE2 "${SOURCE_DIR}/thirdparty/azure-sdk-cpp-for-cpp/fix-illegal-qualified-name-in-member.patch")
+    set(PC bash -c "set -x && (\"${Patch_EXECUTABLE}\" -p1 -N -i \"${PATCH_FILE1}\" &&\
+            \"${Patch_EXECUTABLE}\" -p1 -N -i \"${PATCH_FILE2}\") ||\
+            (\"${Patch_EXECUTABLE}\" -p1 -R --dry-run -i \"${PATCH_FILE1}\" &&\
+            \"${Patch_EXECUTABLE}\" -p1 -R --dry-run -i \"${PATCH_FILE2}\")")
```
Comment (fgerlits) [inline cmake/BundledAzureSdkCpp.cmake:24]: Isn't this backwards?  We want to "dry-run" the reverse patch first, and only run the real patch if the dry-run failed:

```suggestion
    set(PC bash -c "set -x &&\
            (\"${Patch_EXECUTABLE}\" -p1 -R --dry-run -i \"${PATCH_FILE1}\" || \"${Patch_EXECUTABLE}\" -p1 -N -i \"${PATCH_FILE1}\") &&\
            (\"${Patch_EXECUTABLE}\" -p1 -R --dry-run -i \"${PATCH_FILE2}\")" || \"${Patch_EXECUTABLE}\" -p1 -N -i \"${PATCH_FILE2}\") ")
```

File: cmake/CppVersion.cmake [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+function(set_cpp_version)
+    if (MSVC)
+        if ((MSVC_VERSION GREATER "1910") OR (MSVC_VERSION EQUAL "1910"))
+            add_compile_options("/std:c++latest")
+            add_compile_options("/permissive-")
+        else()
+            message(STATUS "The compiler ${CMAKE_CXX_COMPILER} has no C++17 support. Please use a different C++ compiler.")
```
Comment (fgerlits) [inline cmake/CppVersion.cmake:24]: This error message is misleading, as we aren't checking C++17 support directly.
```suggestion
            message(STATUS "The Visual Studio C++ compiler ${CMAKE_CXX_COMPILER} is not supported. Please use Visual Studio 2017 or newer.")
```

File: cmake/BundledAzureSdkCpp.cmake (hunk 2 above)
Comment (szaszm) [reply cmake/BundledAzureSdkCpp.cmake:24]: I copied this from OSSPUUID. The end result is the same, but if you think this is better for some reason, I can apply the change.

Comment (fgerlits) [reply cmake/BundledAzureSdkCpp.cmake:24]: My version ("for each patch: check if the patch has been already applied, and if not, then apply it") makes more sense to me, but this is subjective.  Having read it five more times, I can see now that your (or OSSPUUID's) version of "try to apply both patches, and if at least one of them fails, then check if both patches have been already applied at this point, and if not, then give an error" is logically the same, so I'm OK with either.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -154,7 +111,7 @@ jobs:
         run: |
           PATH %PATH%;C:\Program Files (x86)\Windows Kits\10\bin\10.0.19041.0\x64
           PATH %PATH%;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Roslyn
-          win_build_vs.bat build /2019 /64 /CI /S /A /PDH /K /L /R
+          win_build_vs.bat build /64 /CI /S /A /PDH /K /L /R
```
[the remark this answers is not in the document]
Comment (szaszm) [reply .github/workflows/ci.yml]: Thanks, fixed in 86e3a86

File: cmake/BundledAzureSdkCpp.cmake (hunk 2 above)
Comment (szaszm) [reply cmake/BundledAzureSdkCpp.cmake]: I agree that your version might be easier to follow, so changed it. Thanks.

File: cmake/BundledAzureSdkCpp.cmake [hunk 5]
```diff
@@ -16,8 +16,12 @@
 # under the License.
 
 function(use_bundled_libazure SOURCE_DIR BINARY_DIR)
-    set(PATCH_FILE "${SOURCE_DIR}/thirdparty/azure-sdk-cpp-for-cpp/azure-sdk-for-cpp-old-compiler.patch")
-    set(PC "${Patch_EXECUTABLE}" -R -p1 -s -f --dry-run -i "${PATCH_FILE}" || "${Patch_EXECUTABLE}" -p1 -i "${PATCH_FILE}")
+    set(PATCH_FILE1 "${SOURCE_DIR}/thirdparty/azure-sdk-cpp-for-cpp/azure-sdk-for-cpp-old-compiler.patch")
+    set(PATCH_FILE2 "${SOURCE_DIR}/thirdparty/azure-sdk-cpp-for-cpp/fix-illegal-qualified-name-in-member.patch")
+    set(PC bash -c "set -x &&\
+            (\"${Patch_EXECUTABLE}\" -p1 -R --dry-run -i \"${PATCH_FILE1}\" || \"${Patch_EXECUTABLE}\" -p1 -N -i \"${PATCH_FILE1}\") &&\
+            (\"${Patch_EXECUTABLE}\" -p1 -R --dry-run -i \"${PATCH_FILE2}\")" || \"${Patch_EXECUTABLE}\" -p1 -N -i \"${PATCH_FILE2}\") ")
```
Comment (fgerlits) [inline cmake/BundledAzureSdkCpp.cmake:23]: Sorry, slight f***p on my part -- there was an extra `")` in the middle.
```suggestion
            (\"${Patch_EXECUTABLE}\" -p1 -R --dry-run -i \"${PATCH_FILE2}\" || \"${Patch_EXECUTABLE}\" -p1 -N -i \"${PATCH_FILE2}\") ")
```

Comment (szaszm) [reply cmake/BundledAzureSdkCpp.cmake]: No worries. I didn't notice either. :) Thanks for reacting quickly.

**Review by fgerlits — APPROVED** [review]

File: .gitignore [hunk 6]
```diff
@@ -54,6 +54,7 @@ target
 thirdparty/**/*.o
 thirdparty/**/*.a
 libminifi/test/**/*.a
+libminifi/test/corecomponentstate/
```
Comment (arpadboda) [inline .gitignore:57]: 👍

**Review by arpadboda — APPROVED** [review]
LGTM, please make sure you update this documentation as well when this gets merged:

https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=139627733

**Review by martinzink — APPROVED** [review]

**Pull request closed without merging**